library/std/src/io/buffered/bufwriter.rs

```diff
@@ -327,9 +327,16 @@ impl<W: Write> BufWriter<W> {
     /// ```
     #[unstable(feature = "bufwriter_into_raw_parts", issue = "80690")]
     pub fn into_raw_parts(mut self) -> (W, Result<Vec<u8>, WriterPanicked>) {
-        let buf = mem::take(&mut self.buf);
-        let buf = if !self.panicked { Ok(buf) } else { Err(WriterPanicked { buf }) };
-        (self.inner.take().unwrap(), buf)
+        // SAFETY: forget(self) prevents double dropping inner/buf
+        unsafe {
+            let inner = ptr::read(&mut self.inner);
+            let buf = ptr::read(&mut self.buf);
```

Comment:
The `mem::take` approach should still work for the buffer, no? It replaces it with an empty iterator which doesn't have any allocation so there's no issue leaking it.

Reply:
It would yeah, I figured since the unsafe is there anyway it would be clear enough to read them both the same way, and saves zeroing out the buf in self


I can use take again if that's preferred though

Reply:
I assume that the zeroing cost is tiny compared to whatever IO is done with the writer. Using take keeps the unsafe block as small as possible.

You can shrink it further by moving the following lines out of the block, they're not unsafe after all. All that's needed is one `let inner =  unsafe { ptr::read() }`;